janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -603,23 +605,27 @@ private static Query numericQuery(String key, Cmp relation, Number value) {
     }
 
     // adapted from SolrIndex
-    private List<String> customTokenize(Analyzer analyzer, String fieldName, String value) {
-        final List<String> terms = new ArrayList<>();
+    private List<List<String>> customTokenize(Analyzer analyzer, String fieldName, String value) {
+        Map<Integer, List<String>> stemsByOffset = new HashMap<>();
         try (CachingTokenFilter stream = new CachingTokenFilter(analyzer.tokenStream(fieldName, value))) {
+            final OffsetAttribute offsetAtt = stream.getAttribute(OffsetAttribute.class);
             final TermToBytesRefAttribute termAtt = stream.getAttribute(TermToBytesRefAttribute.class);
             stream.reset();
             while (stream.incrementToken()) {
-                terms.add(termAtt.getBytesRef().utf8ToString());
+                int offset = offsetAtt.startOffset();
+                String stem = termAtt.getBytesRef().utf8ToString();
+                stemsByOffset.putIfAbsent(offset, new ArrayList<>());
```

Comment:
(nitpick) 
So we are executing `new ArrayList<>()` even if it isn't absent. Generally it isn't a big deal because GC handles those objects easily but I am not sure if compilator is able to optimize that (I think it isn't).
You can either create a static function which returns a new array list (i.e. it is executed only if it is absent) or your can just return your list and create a new one if it is empty.
Something like:
```
List<String> stemList = stemsByOffset.get(offset);
if(stemList == null){
  stemList = new ArrayList<>();
  stemsByOffset.put(offset, stemList);
}
stemList.add(stem);
```

Reply:
It makes sense, thanks.
Done.